File: go/parser.go
package promptmd

import (
	"bufio"
	"os"
	"strings"

	"gopkg.in/yaml.v2"
)

type Prompt struct {
	Role    string
	Content string
}

type PromptFile struct {
	FrontMatter map[string]interface{}
	Prompts     []Prompt
}

func ParsePromptFile(filePath string) (*PromptFile, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var frontMatterLines []string
	var prompts []Prompt
	var currentRole string
	var currentContent strings.Builder
	inFrontMatter := false
	inPrompt := false

	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "---" {
			if !inFrontMatter {
				inFrontMatter = true
			} else {
				inFrontMatter = false
				inPrompt = true
			}
			continue
		} else {
			if !inFrontMatter && !inPrompt {
				inPrompt = true
			}
		}

		if inFrontMatter {
			frontMatterLines = append(frontMatterLines, line)
		} else if inPrompt {
			if strings.HasSuffix(line, ":") && (strings.HasPrefix(line, "user") || strings.HasPrefix(line, "system") || strings.HasPrefix(line, "assistant")) {
				if currentRole != "" {
					prompts = append(prompts, Prompt{
						Role:    currentRole,
						Content: currentContent.String(),
					})
					currentContent.Reset()
				}
				currentRole = strings.TrimSuffix(line, ":")
			} else {
				currentContent.WriteString(line + "\n")
			}
		}
	}

	if currentRole != "" {
		prompts = append(prompts, Prompt{
			Role:    currentRole,
			Content: currentContent.String(),
		})
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	frontMatter := make(map[string]interface{})
	if len(frontMatterLines) > 0 {
		err = yaml.Unmarshal([]byte(strings.Join(frontMatterLines, "\n")), &frontMatter)
		if err != nil {
			return nil, err
		}
	}

	return &PromptFile{
		FrontMatter: frontMatter,
		Prompts:     prompts,
	}, nil
}

File: ts/parser.ts
import * as fs from 'fs';
import * as yaml from 'yaml';

interface Prompt {
  role: string;
  content: string;
}

interface PromptFile {
  frontMatter: Record<string, any>;
  prompts: Prompt[];
}

function parsePromptFile(filePath: string): Promise<PromptFile> {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, 'utf8', (err, data) => {
      if (err) {
        return reject(err);
      }

      const lines = data.split('\n');
      let frontMatterLines: string[] = [];
      let prompts: Prompt[] = [];
      let currentRole: string = '';
      let currentContent: string[] = [];
      let inFrontMatter = false;
      let inPrompt = false;

      for (const line of lines) {
        if (line.trim() === '---') {
          if (!inFrontMatter) {
            inFrontMatter = true;
          } else {
            inFrontMatter = false;
            inPrompt = true;
          }
          continue;
        } else {
          if (!inFrontMatter && !inPrompt) {
            inPrompt = true;
          }
        }

        if (inFrontMatter) {
          frontMatterLines.push(line);
        } else if (inPrompt) {
          if (line.endsWith(':') && (line.startsWith('user') || line.startsWith('system') || line.startsWith('assistant'))) {
            if (currentRole) {
              prompts.push({
                role: currentRole,
                content: currentContent.join('\n'),
              });
              currentContent = [];
            }
            currentRole = line.slice(0, -1);
          } else {
            currentContent.push(line);
          }
        }
      }

      if (currentRole) {
        prompts.push({
          role: currentRole,
          content: currentContent.join('\n'),
        });
      }

      let frontMatter: Record<string, any> = {};
      if (frontMatterLines.length > 0) {
        try {
          frontMatter = yaml.parse(frontMatterLines.join('\n'));
        } catch (err) {
          return reject(err);
        }
      }

      resolve({
        frontMatter,
        prompts,
      });
    });
  });
}

export { Prompt, PromptFile, parsePromptFile };
